python/tvm/relax/training/loss.py

```diff
@@ -0,0 +1,240 @@
+from ..op.nn import log_softmax, nll_loss
+
+
+__all__ = ["L1Loss", "MSELoss", "CrossEntropyLoss"]
+
+
+def _create_param_var(param: Union[Var, StructInfo], param_name) -> Var:
+    if isinstance(param, StructInfo):
+        param = Var(param_name, param)
+    assert isinstance(param, Var)
+    return param
+
+
+class Loss:
+    """Base class of all loss.
+
+    Parameters
+    ----------
+    loss_name : str
+        The name of the loss function.
+
+    reduction : str
+        The reduction method to apply to output. Can be "mean", "sum" or "none".
+
+        none : no reduction will be applied,
+        mean : the sum of the output will be divided by the batch_size,
+        sum : the output will be summed.
+    """
+
+    reduction: str
+    loss_name: str
+
+    def __init__(self, loss_name: str, reduction: str = "mean") -> None:
+        self.loss_name = loss_name
+        self.reduction = reduction
+
+        valid_reductions = ["mean", "sum", "none"]
+
+        if self.reduction not in valid_reductions:
+            raise ValueError("Reduction can only be one of these values: ", valid_reductions)
+
+    def _with_reduction(self, expr: Expr):
+        """Add a reduction to the final loss.
+
+        Parameters
+        ----------
+        expr : Expr
+            The loss expr.
+        """
+        if self.reduction == "sum":
+            expr = sum(expr)
+        elif self.reduction == "mean":
+            expr = mean(expr)
+        else:
+            assert self.reduction == "none"
+        return expr
+
+
+class L1Loss(Loss):
+    """Mean element-wise absolute value difference.
+
+    Parameters
+    ----------
+    reduction : str
+        See the doc of Loss.
+    """
+
+    def __init__(self, reduction: str = "mean") -> None:
+        super().__init__("l1_loss", reduction)
+
+    def __call__(
+        self,
+        predictions: Union[Var, StructInfo],
+        targets: Union[Var, StructInfo],
+    ) -> Function:
+        """Get the relax function of L1Loss. If the parameters are
+        struct info, it will create corresponding variables.
+
+        Parameters
+        ----------
+        predictions : Union[Var, StructInfo]
+            The predictions of the model in the calculation of loss.
+        targets : Union[Var, StructInfo]
+            The ground truth in the calculation of loss.
+
+        Returns
+        ----------
+        The relax function of L1Loss with the loss name as its global symbol.
+        """
+        bb = relax.BlockBuilder()
+
+        predictions = _create_param_var(predictions, "predictions")
+        targets = _create_param_var(targets, "targets")
+
+        with bb.function(self.loss_name, [predictions, targets]):
+            with bb.dataflow():
+                lv = abs(subtract(predictions, targets))
+                loss = bb.emit_output(self._with_reduction(lv))
+            bb.emit_func_output(loss)
+
+        return bb.get()[self.loss_name].with_attr("global_symbol", self.loss_name)
```

Comment:
Is global symbol necessary here?

Reply:
The loss can be called by others and I think it's good we annotate it with a gsymbol to make it a complete relax Function